src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt

```diff
@@ -230,6 +232,83 @@ internal val Class<*>.allDeclaredFieldWithSuperclasses get(): List<Field> {
+ * the return value of the callback determines, what object should be traversed further
+ * (it can be something different from `fieldValue`).
+ *
+ * Parameters of [callback]:
+ * - `fieldOwner`: either `Any` - some class instance that owns the field, or `Class<*>` is case if `field` is static.
+ * - `field`: the field that currently is traversed.
+ * - `fieldValue`: the value that was read from the field.
+ */
+internal fun traverseObjectHierarchy(obj: Any, treatArrayElementsAsFields: Boolean = true, callback: (Any, Field, Any?) -> Any?) {
+    val queue = ArrayDeque<Any>()
+    val visitedObjects = Collections.newSetFromMap<Any>(IdentityHashMap())
+
+    queue.add(obj)
+    visitedObjects.add(obj)
+
+    while (queue.isNotEmpty()) {
+        val currentObj = queue.removeFirst()
+        val fields =
+            if (currentObj is Class<*>) emptyList<Field>()
+            else if (currentObj.javaClass.isArray && treatArrayElementsAsFields) {
+                // We do not traverse the actual fields of an array,
+                // instead the elements of the array are treated as its field.
+                // But note that primitive arrays are skipped completely.
+                if (currentObj is Array<*>) {
+                    currentObj.forEach { element ->
+                        element?.let { queue.add(it) }
+                    }
+                }
+                emptyList<Field>()
+            }
+            else if (
+                (currentObj is AtomicReferenceArray<*> || currentObj is kotlinx.atomicfu.AtomicArray<*>) &&
+                treatArrayElementsAsFields
+            ) {
+                // Do the same as before, but for atomic arrays (with non-primitive types)
+                val getElementAt = currentObj.javaClass.getMethod("get", Int::class.javaPrimitiveType)
+                val length = (
+                    if (currentObj is AtomicReferenceArray<*>) currentObj.javaClass.getMethod("length").invoke(currentObj)
+                    else currentObj.javaClass.getMethod("getSize").invoke(currentObj)
+                ) as Int
+
+                for (index in 0..length - 1) {
+                    getElementAt.invoke(currentObj, index)?.let { queue.add(it) }
+                }
+                emptyList<Field>()
+            }
+            else currentObj.javaClass.allDeclaredFieldWithSuperclasses
+
+        for (f in fields) {
+            // We wrap an unsafe read into `runCatching` to hande `UnsupportedOperationException`,
+            // which can be thrown, for instance, when attempting to read a field of
+            // a hidden class (starting from Java 15).
+            val result = runCatching { readField(currentObj, f) }
+            if (result.isFailure) continue // do not pass fields to the user, that are non-readable by Unsafe
+
+            val fieldValue = result.getOrNull()
+            val nextObject = callback(currentObj, f, fieldValue)
```

Comment:
Why the semantics of `callback` is to return `nextObject`, instead of just boolean `shouldTraverse` flag as before?

Reply:
In order to allow skipping through fields

Reply:
changed back to return `true`/`false`)